docs/design/features/MarshalDirective.md

```diff
@@ -0,0 +1,206 @@
+
+6) Once the export returns control to the IL Stub the marshaling logic cleans up and ensures any returned data is marshaled back out to the calling function.
+
+An example of how the snippet above could be transformed is below. The example is using the proposed API in this document as well as the [C# function pointer proposal][csharp_fptr_link]. Observe points (C), the new attribute, (D), a `partial` class, and (E) the stub with additional `IntPtr` argument that will be called in lieu of a runtime generated IL Stub.
+
+``` CSharp
+[DllImportAttribute("Kernel32.dll")]
+/* C */ [MarshalDirectiveAttribute(
+        Type = MarshalDirectiveType.UserDefinedStub,
+        UserFunctionType = typeof(Stubs),
+        UserDefinedStub = nameof(QueryPerformanceCounterStub))]
+/* A */ extern static bool QueryPerformanceCounter(out long lpPerformanceCount);
+...
+long count;
+/* B */ QueryPerformanceCounter(out count);
+
+/* D */  class partial Stubs
+{
+    /* E */ public static bool QueryPerformanceCounterStub(IntPtr fptr, out long lpPerformanceCount)
+    {
+        unsafe
+        {
+            long result = 0;
+            var tfptr = (delegate* stdcall<long*, int>)fptr;
+            bool success = tfptr(&result) != 0;
+            lpPerformanceCount = result;
+            return success;
+        }
+    }
+}
+```
+
+The contents of the above stub are trivial and in a `Release` build wouldn't even exist. More complicated examples may require [GC handle pinning](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.gchandle) or native memory allocation. These complicated examples could be hand authored or generated through a third-party tool.
+
+In this system it is not defined how marshaling of specific types would be performed. The built-in runtime has complex rules for some types and it is these rules that once shipped become the defacto standard - often times regardless if the behavior is a bug or not. The design here is not concerned with how the arguments go from a managed to unamanged environment.
+
+### Proposed API
+``` CSharp
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// User defined marshaling directive
+    /// </summary>
+    public enum MarshalDirectiveType
+    {
+        /// <summary>
+        /// The attribute does nothing.
+        /// </summary>
+        Default = 0,
+
+        /// <summary>
+        /// The user can supply a managed function that will be called in-place
+        /// of a generated function stub.
+        /// </summary>
+        /// <seealso cref="MarshalDirectiveAttribute.UserDefinedStub"/>
+        UserDefinedStub,
+    }
+
+    /// <summary>
+    /// Attribute used to define how marshaling stubs should be defined for
+    /// the associated managed function declaration.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = false)]
+    public sealed class MarshalDirectiveAttribute : Attribute
```

Comment:
This looks like ManagedToNativeComInteropStubAttribute just for regular PInvokes.

Reply:
IIRC, ManagedToNativeComInteropStubAttribute was done just for COM and not for regular PInvokes because of there is no alternative with built in COM to provide custom marshaling. But it is not the case for regular PInvoke - you can easily write the marshaling yourself.

Reply:
Yes.